---
layout: post
title: 出场率比较高的一道多线程安全面试题
categories: [基础]
excerpt: 出场率比较高的一道多线程安全面试题
tags: java， 理论  
---

这个问题是 Java 程序员面试经常会遇到的吧。

工作一两年的应该都知道 ArrayList 是线程不安全的，要使用线程安全的就使用 Vector，这也是各种 Java 面试宝典里面所提及的，可能很多工作好几年的程序员都停留在这个知识面上。

先说说为什么 ArrayList 是线程不安全的吧，来看以下的代码。

```
/**
 * 微信公众号：Java技术栈
 */
public class TestArrayList {

	private static List<Integer> list = new ArrayList<>();

	public static void main(String[] args) throws InterruptedException {
		for (int i = 0; i < 10; i++) {
			testList();
			list.clear();
		}
	}

	private static void testList() throws InterruptedException {
		Runnable runnable = () -> {
			for (int i = 0; i < 10000; i++) {
				list.add(i);
			}
		};

		Thread t1 = new Thread(runnable);
		Thread t2 = new Thread(runnable);
		Thread t3 = new Thread(runnable);

		t1.start();
		t2.start();
		t3.start();

		t1.join();
		t2.join();
		t3.join();

		System.out.println(list.size());
	}

}
```

这是它的输出结果，我们期望的结果应该都是：30000，然后并不是，这就是传说中的多线程并发问题了。

```
Exception in thread "Thread-1" java.lang.ArrayIndexOutOfBoundsException: 15786
	at java.base/java.util.ArrayList.add(ArrayList.java:468)
	at java.base/java.util.ArrayList.add(ArrayList.java:480)
	at com.test.thread.TestArrayList.lambda$testList$0(TestArrayList.java:23)
	at java.base/java.lang.Thread.run(Thread.java:844)
20332
16100
14941
23749
15631
22118
27417
30000
28691
27843
```

#### 现象分析

从以上结果可以总结出 ArrayList 在并发情况下会出现的几种现象。

**1、发生 ArrayIndexOutOfBoundsException 异常；**

```
private void add(E e, Object[] elementData, int s) {
    if (s == elementData.length)
        elementData = grow();
    elementData[s] = e;
    size = s + 1;
}
```

定位到异常所在源代码，毫无疑问，问题是出现在多线程并发访问下，由于没有同步锁的保护，造成了 ArrayList 扩容不一致的问题。


**2、程序正常运行，输出了少于实际容量的大小；**

这个也是多线程并发赋值时，对同一个数组索引位置进行了赋值，所以出现少于预期大小的情况。

**3、程序正常运行，输出了预期容量的大小；**

这是正常运行结果，未发生多线程安全问题，但这是不确定性的，不是每次都会达到正常预期的。

#### 解决方案

既然这样，那么在高并发情况下，使用什么样的列表集合保护线程安全呢？回到文章最开始的地方，使用 Vector，还有别的吗？当然有，篇幅有限，请各位看官期待后续文章。

另外，像 HashMap, HashSet 等都有类似多线程安全问题，在多线程并发环境下避免使用这种集合。

